apps/src/weblab/WebLab.js

```diff
@@ -297,11 +297,19 @@ WebLab.prototype.init = function(config) {
     document.getElementById(config.containerId)
   );
 
-  window.onbeforeunload = evt => {
-    if (project.hasOwnerChangedProject()) {
-      return weblabMsg.confirmExitWithUnsavedChanges();
-    }
-  };
+  window.addEventListener('beforeunload', this.beforeUnload.bind(this));
+};
+
+WebLab.prototype.beforeUnload = function(event) {
+  if (project.hasOwnerChangedProject()) {
+    // Manually trigger an autosave instead of waiting for the next autosave.
+    project.autosave();
+
+    event.preventDefault();
+    event.returnValue = weblabMsg.confirmExitWithUnsavedChanges();
+  } else {
+    delete event.returnValue;
```

Comment:
What does this do?

Reply:
the event's `returnValue` indicates whether or not the default action for that event has been prevented ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/Event/returnValue)), and removing `returnValue` should guarantee that the browser unloads ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload#example)). it's basically just a safety precaution to make sure the browser unloads.

now that i'm reading about this again i'm realizing that line 309 is outdated:
```
event.returnValue = weblabMsg.confirmExitWithUnsavedChanges();

// should be
e.returnValue = '';
```
some browsers used to display the `returnValue` in the resulting alert dialog, but it doesn't look like any modern browsers still do that